readme.md
# Google2FA

[![Latest Stable Version](https://img.shields.io/packagist/v/rdtvaacar/google2fa.svg?style=flat-square)](https://packagist.org/packages/rdtvaacar/google2fa) [![License](https://img.shields.io/badge/license-BSD_3_Clause-brightgreen.svg?style=flat-square)](LICENSE) [![Downloads](https://img.shields.io/packagist/dt/rdtvaacar/google2fa.svg?style=flat-square)](https://packagist.org/packages/rdtvaacar/google2fa)

### Google Two-Factor Authentication for PHP Package

Google2FA is a PHP implementation of the Google Two-Factor Authentication Module, supporting the HMAC-Based One-time Password (HOTP) algorithm specified in [RFC 4226](https://tools.ietf.org/html/rfc4226) and the Time-based One-time Password (TOTP) algorithm specified in [RFC 6238](https://tools.ietf.org/html/rfc6238).

This package is agnostic, but also supports the Laravel Framework.

## Requirements

- PHP 5.3.7+

## Compatibility

You don't need Laravel to use it, but it's compatible with

- Laravel 4.1+
- Laravel 5+

## Installing

Use Composer to install it:

    composer require rdtvaacar/google2fa

If you prefer inline QRCodes instead of a Google generated url, you'll need to install [BaconQrCode](https://github.com/Bacon/BaconQrCode):
  
    composer require "bacon/bacon-qr-code":"~1.0"

## Installing on Laravel

Add the Service Provider and Facade alias to your `app/config/app.php` (Laravel 4.x) or `config/app.php` (Laravel 5.x):

    Rdtvaacar\Google2FA\Vendor\Laravel\ServiceProvider::class,

    'Google2FA' => Rdtvaacar\Google2FA\Vendor\Laravel\Facade::class,

## Using It

#### Instantiate it directly

```
use Rdtvaacar\Google2FA\Google2FA;

$google2fa = new Google2FA();

return $google2fa->generateSecretKey();
```

#### In Laravel you can use the IoC Container and the contract

```
$google2fa = app()->make('Rdtvaacar\Google2FA\Contracts\Google2FA');

return $google2fa->generateSecretKey();
```

#### Or Method Injection, in Laravel 5

```
use Rdtvaacar\Google2FA\Contracts\Google2FA;

class WelcomeController extends Controller 
{
	public function generateKey(Google2FA $google2fa)
	{
		return $google2fa->generateSecretKey();
	}
}
```

#### Or the Facade

```
return Google2FA::generateSecretKey();
```

## How To Generate And Use Two Factor Authentication

Generate a secret key for your user and save it:

    $user = User::find(1);

    $user->google2fa_secret = Google2FA::generateSecretKey();

    $user->save();

Show the QR code to your user:

    $google2fa_url = Google2FA::getQRCodeGoogleUrl(
    	'YourCompany',
    	$user->email,
    	$user->google2fa_secret
    );

	{{ HTML::image($google2fa_url) }}

And they should see and scan the QR code to their applications:

![QRCode](https://chart.googleapis.com/chart?chs=200x200&chld=M|0&cht=qr&chl=otpauth%3A%2F%2Ftotp%2FPragmaRX%3Aacr%2Bpragmarx%40antoniocarlosribeiro.com%3Fsecret%3DADUMJO5634NPDEKW%26issuer%3DPragmaRX)

And to verify, you just have to:

	$secret = Input::get('secret');

    $valid = Google2FA::verifyKey($user->google2fa_secret, $secret);

## Server Time

It's really important that you keep your server time in sync with some NTP server, on Ubuntu you can add this to the crontab:

    ntpdate ntp.ubuntu.com

## Using a Bigger and Prefixing the Secret Key

Although the probability of collision of a 16 bytes (128 bits) random string is very low, you can harden it by:
 
#### Use a bigger key

    $secretKey = $google2fa->generateSecretKey(32); // defaults to 16 bytes

#### Prefix it

    $secretKey = $google2fa->generateSecretKey(16, $userId);

## Demos

Here's a demo app showing how to use Google2FA: [google2fa-example](https://github.com/antonioribeiro/google2fa-example).

You can scan the QR code on [this page](https://antoniocarlosribeiro.com/technology/google2fa) with a Google Authenticator app and view the code changing (almost) in real time.

## Google Authenticator Apps:

To use the two factor authentication, your user will have to install a Google Authenticator compatible app, those are some of the currently available:

* [Authy for iOS, Android, Chrome, OS X](https://www.authy.com/)
* [FreeOTP for iOS, Android and Peeble](https://fedorahosted.org/freeotp/)
* [FreeOTP for iOS, Android and Peeble](https://www.toopher.com/)
* [Google Authenticator for iOS](http://itunes.apple.com/us/app/google-authenticator/id388497605?mt=8")
* [Google Authenticator for Android](https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2")
* [Google Authenticator for Blackberry](https://m.google.com/authenticator")
* [Google Authenticator (port) on Windows app store](http://apps.microsoft.com/windows/en-us/app/google-authenticator/7ea6de74-dddb-47df-92cb-40afac4d38bb")
* [Microsoft Authenticator for Windows Phone](https://www.microsoft.com/en-us/store/apps/authenticator/9wzdncrfj3rj)
* [1Password for iOS, Android, OSX, Windows](https://1password.com)

## Tests

The package tests were written with [phpspec](http://www.phpspec.net/en/latest/).

## Author

[Antonio Carlos Ribeiro](http://twitter.com/iantonioribeiro)

## License

Google2FA is licensed under the BSD 3-Clause License - see the `LICENSE` file for details

## Contributing

Pull requests and issues are more than welcome.


changelog.md
## Change Log

## [1.0.0] - 2016-07-17
### Changed
- Package bacon/bacon-qr-code was moved to "suggest". 

## [0.8.1] - 2016-07-17
### Fixed
- Allow paragonie/random_compat ~1.4|~2.0.

## [0.8.0] - 2016-07-17
### Changed
- Bumped christian-riesen/base32 to ~1.3
- Use paragonie/random_compat to generate cryptographically secure random secret keys
- Readme improvements
- Drop simple-qrcode in favor of bacon/bacon-qr-code 
- Fix tavis setup for phpspec, PHP 7, hhvm and improve cache

## [0.7.0] - 2015-11-07
### Changed
- Fixed URL generation for QRCodes
- Avoid time attacks

## [0.2.0] - 2015-02-19
### Changed
- Laravel 5 compatibility.

## [0.1.0] - 2014-07-06
### Added
- First version.


src/Google2FA.php
<?php

namespace Rdtvaacar\Google2FA;

/**
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * PHP Google two-factor authentication module.
 *
 * See http://www.idontplaydarts.com/2011/07/google-totp-two-factor-authentication-for-php/
 * for more details
 *
 * @author Phil (Orginal author of this class)
 *
 * Changes have been made in the original class to remove all static methods and, also,
 * provide some other methods.
 *
 * @package    Google2FA
 * @author     Antonio Carlos Ribeiro @ Rdtvaacar
 **/


use BaconQrCode\Writer;
use BaconQrCode\Renderer\Image\Png;
use Rdtvaacar\Google2FA\Support\Url;
use Rdtvaacar\Google2FA\Exceptions\InvalidCharactersException;
use Rdtvaacar\Google2FA\Exceptions\SecretKeyTooShortException;
use Rdtvaacar\Google2FA\Contracts\Google2FA as Google2FAContract;
use Tuupola\Base32;

class Google2FA implements Google2FAContract
{
    /**
     * Interval between key regeneration
     */
    const KEY_REGENERATION = 30;

    /**
     * Length of the Token generated.
     *
     */
    const OPT_LENGTH = 6;

    /**
     * Characters valid for Base 32.
     *
     */
    const VALID_FOR_B32 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

    /**
     * Generate a digit secret key in base32 format.
     *
     * @param int $length
     * @return string
     */
    public function generateSecretKey($length = 16, $prefix = '')
    {
        $b32 = "234567QWERTYUIOPASDFGHJKLZXCVBNM";

        $secret = $prefix ? $this->toBase32($prefix) : '';

        for ($i = 0; $i < $length; $i++) {
            $secret .= $b32[$this->getRandomNumber()];
        }

        $this->validateSecret($secret);

        return $secret;
    }

    /**
     * Returns the current Unix Timestamp divided by the KEY_REGENERATION
     * period.
     *
     * @return integer
     **/
    public function getTimestamp()
    {
        return floor(microtime(true) / static::KEY_REGENERATION);
    }

    /**
     * Decodes a base32 string into a binary string.
     *
     * @param string $b32
     * @return integer
     * @throws InvalidCharactersException
     */
    public function base32Decode($b32)
    {
        $b32 = strtoupper($b32);
        $this->validateSecret($b32);
        $base32 = new Base32();
        return $base32->decode($b32);
    }

    /**
     * Takes the secret key and the timestamp and returns the one time
     * password.
     *
     * @param string $key - Secret key in binary form.
     * @param integer $counter - Timestamp as returned by getTimestamp.
     * @return string
     * @throws SecretKeyTooShortException
     */
    public function oathHotp($key, $counter)
    {
        if (strlen($key) < 8) {
            throw new SecretKeyTooShortException();
        }

        // Counter must be 64-bit int
        $bin_counter = pack('N*', 0, $counter);

        $hash = hash_hmac('sha1', $bin_counter, $key, true);

        return str_pad($this->oathTruncate($hash), static::OPT_LENGTH, '0', STR_PAD_LEFT);
    }

    /**
     * Get the current one time password for a key.
     *
     * @param string $initalizationKey
     * @return string
     * @throws InvalidCharactersException
     * @throws SecretKeyTooShortException
     */
    public function getCurrentOtp($initalizationKey)
    {
        $timestamp = $this->getTimestamp();

        $secretKey = $this->base32Decode($initalizationKey);

        return $this->oathHotp($secretKey, $timestamp);
    }

    /**
     * Verifies a user inputted key against the current timestamp. Checks $window
     * keys either side of the timestamp.
     *
     * @param string $b32seed
     * @param string $key - User specified key
     * @param integer $window
     * @param boolean $useTimeStamp
     * @return boolean
     **/
    public function verifyKey($b32seed, $key, $window = 4, $useTimeStamp = true)
    {
        $timeStamp = $this->getTimestamp();

        if ($useTimeStamp !== true) {
            $timeStamp = (int)$useTimeStamp;
        }

        $binarySeed = $this->base32Decode($b32seed);

        for ($ts = $timeStamp - $window; $ts <= $timeStamp + $window; $ts++) {
            if (hash_equals($this->oathHotp($binarySeed, $ts), $key)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Extracts the OTP from the SHA1 hash.
     *
     * @param string $hash
     * @return integer
     **/
    public function oathTruncate($hash)
    {
        $offset = ord($hash[19]) & 0xf;
        $temp   = unpack('N', substr($hash, $offset, 4));
        return substr($temp[1] & 0x7fffffff, -static::OPT_LENGTH);
    }

    /**
     * Remove invalid chars from a base 32 string.
     *
     * @param $string
     * @return mixed
     */
    public function removeInvalidChars($string)
    {
        return preg_replace('/[^' . static::VALID_FOR_B32 . ']/', '', $string);
    }

    /**
     * Creates a Google QR code url.
     *
     * @param string $company
     * @param string $holder
     * @param string $secret
     * @param integer $size
     * @return string
     */
    public function getQRCodeGoogleUrl($company, $holder, $secret, $size = 200)
    {
        $url = $this->getQRCodeUrl($company, $holder, $secret);

        return Url::generateGoogleQRCodeUrl('https://chart.googleapis.com/', 'chart', 'chs=' . $size . 'x' . $size . '&chld=M|0&cht=qr&chl=', $url);
    }

    /**
     * Generates a QR code data url to display inline.
     *
     * @param string $company
     * @param string $holder
     * @param string $secret
     * @param integer $size
     * @param string $encoding Default to UTF-8
     * @return string
     */
    public function getQRCodeInline($company, $holder, $secret, $size = 200, $encoding = 'utf-8')
    {
        $url = $this->getQRCodeUrl($company, $holder, $secret);

        $renderer = new Png();
        $renderer->setWidth($size);
        $renderer->setHeight($size);

        $writer = new Writer($renderer);
        $data   = $writer->writeString($url, $encoding);

        return 'data:image/png;base64,' . base64_encode($data);
    }

    /**
     * Creates a QR code url.
     *
     * @param $company
     * @param $holder
     * @param $secret
     * @return string
     */
    public function getQRCodeUrl($company, $holder, $secret)
    {
        return 'otpauth://totp/' . $company . ':' . $holder . '?secret=' . $secret . '&issuer=' . $company . '';
    }

    /**
     * Get a random number.
     *
     * @param $from
     * @param $to
     * @return int
     */
    private function getRandomNumber($from = 0, $to = 31)
    {
        return random_int($from, $to);
    }

    /**
     * Validate the secret.
     *
     * @param $b32
     * @throws InvalidCharactersException
     */
    private function validateSecret($b32)
    {
        if (!preg_match('/^[' . static::VALID_FOR_B32 . ']+$/', $b32, $match)) {
            throw new InvalidCharactersException();
        }
    }

    /**
     * Encode a string to Base32.
     *
     * @param $string
     * @return mixed
     */
    public function toBase32($string)
    {
        $base32  = new Base32();
        $encoded = $base32->encode($string);
        return str_replace('=', '', $encoded);
    }
}
